examples/DataApi-Voting-poll/template/modules/shared_data/src/main/scala/com/my/currency/shared_data/Main.scala

```diff
@@ -0,0 +1,86 @@
+package com.my.currency.shared_data
+
+import cats.conversions.all.autoWidenFunctor
+import cats.data.NonEmptyList
+import cats.effect.IO
+import cats.implicits.catsSyntaxValidatedIdBinCompat0
+import com.my.currency.shared_data.Utils.getLastMetagraphIncrementalSnapshotInfo
+import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
+import io.circe.{Decoder, Encoder}
+import org.tessellation.currency.dataApplication.dataApplication.DataApplicationValidationErrorOr
+import org.tessellation.currency.dataApplication.{DataApplicationValidationError, DataState, L0NodeContext, L1NodeContext}
+import org.tessellation.security.signature.Signed
+import com.my.currency.shared_data.combiners.Combiners.{combineCreatePoll, combineVoteInPoll}
+import com.my.currency.shared_data.errors.Errors.CouldNotGetLatestCurrencySnapshot
+import com.my.currency.shared_data.validations.Validations.{createPollValidations, createPollValidationsWithSignature, voteInPollValidations, voteInPollValidationsWithSignature}
+import com.my.currency.shared_data.types.Types.{CreatePoll, PollUpdate, VoteCalculatedState, VoteInPoll, VoteStateOnChain}
+import org.slf4j.LoggerFactory
+import org.tessellation.security.SecurityProvider
+
+object Main {
+
+  private val logger = LoggerFactory.getLogger("Main")
+
+  def validateUpdate(update: PollUpdate)(implicit context: L1NodeContext[IO]): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val lastCurrencySnapshot = context.getLastCurrencySnapshot
+
+    lastCurrencySnapshot.map(_.get.ordinal).flatMap { lastSnapshotOrdinal =>
+      update match {
+        case poll: CreatePoll =>
+          createPollValidations(poll, None, Some(lastSnapshotOrdinal))
+        case voteInPoll: VoteInPoll =>
+          getLastMetagraphIncrementalSnapshotInfo(Right(context)) match {
+            case Some(snapshotInfo) => voteInPollValidations(voteInPoll, None, Some(lastSnapshotOrdinal), snapshotInfo)
+            case None => IO(CouldNotGetLatestCurrencySnapshot.asInstanceOf[DataApplicationValidationError].invalidNec)
+          }
+      }
+    }
+  }
+
+  def validateData(state: DataState[VoteStateOnChain, VoteCalculatedState], updates: NonEmptyList[Signed[PollUpdate]])(implicit context: L0NodeContext[IO]): IO[DataApplicationValidationErrorOr[Unit]] = {
+    implicit val sp: SecurityProvider[IO] = context.securityProvider
+    updates.traverse { signedUpdate =>
+      signedUpdate.value match {
+        case poll: CreatePoll =>
+          createPollValidationsWithSignature(poll, signedUpdate.proofs, state)
+        case voteInPoll: VoteInPoll =>
+          getLastMetagraphIncrementalSnapshotInfo(Left(context)) match {
+            case Some(snapshotInfo) => voteInPollValidationsWithSignature(voteInPoll, signedUpdate.proofs, state, snapshotInfo)
+            case None => IO(CouldNotGetLatestCurrencySnapshot.asInstanceOf[DataApplicationValidationError].invalidNec)
```

Comment:
1. The result is a constant so it can be lifted into the effect with `pure`
2. `.asInstanceOf` ? Is this really necessary? If it is, that raises more concerns about the way the error class was defined.